File: C++/0121-best-time-to-buy-and-sell-stock.cpp
class Solution
{
public:
    int maxProfit(vector<int> &prices)
    {
        int min = 10000;
        int max_profit = 0;
        for (int i = 0; i < prices.size(); i++)
        {
            if (min > prices[i])
            {
                min = prices[i];
            }
            else
            {
                max_profit = max(max_profit, prices[i] - min);
            }
        }
        return max_profit;
    }
};

File: Python/0121-best-time-to-buy-and-sell-stock.py
# time complexity: O(n)
# space complexity: O(1)
from typing import List


class Solution(object):
    def maxProfit(self, prices: List[int]) -> int:
        profit = 0
        smallestPirce = float("inf")
        for price in prices:
            smallestPirce = min(smallestPirce, price)
            profit = max(profit, price - smallestPirce)
        return profit


prices = [7, 1, 5, 3, 6, 4]
print(Solution().maxProfit(prices))
prices = [7, 6, 4, 3, 1]
print(Solution().maxProfit(prices))
